src/Server/TCPHandler.cpp

```diff
@@ -243,7 +241,7 @@ void TCPHandler::runImpl()
                 state.logs_queue = std::make_shared<InternalTextLogsQueue>();
                 state.logs_queue->max_priority = Poco::Logger::parseLevel(client_logs_level.toString());
                 CurrentThread::attachInternalTextLogsQueue(state.logs_queue, client_logs_level);
-                CurrentThread::setFatalErrorCallback([this, &fatal_error_mutex]
+                CurrentThread::setFatalErrorCallback([this]
```

Comment:
Will fix the problem too, but this will leave error callback from previous query, so something like this may be better, thoughts?

```patch
diff --git a/src/Server/TCPHandler.cpp b/src/Server/TCPHandler.cpp
index cdf1838c06..12d38d0659 100644
--- a/src/Server/TCPHandler.cpp
+++ b/src/Server/TCPHandler.cpp
@@ -206,8 +206,12 @@ void TCPHandler::runImpl()
 
         try
         {
-            /// If a user passed query-local timeouts, reset socket to initial state at the end of the query
-            SCOPE_EXIT({state.timeout_setter.reset();});
+            SCOPE_EXIT({
+                /// If a user passed query-local timeouts, reset socket to initial state at the end of the query
+                state.timeout_setter.reset();
+                /// Reset callback to avoid using callback from the previous query.
+                CurrentThread::setFatalErrorCallback({});
+            });
 
             /** If Query - process it. If Ping or Cancel - go back to the beginning.
              *  There may come settings for a separate query that modify `query_context`.
```